msal_extensions/_windows.py

```diff
@@ -0,0 +1,173 @@
+# https://stackoverflow.com/questions/463832/using-dpapi-with-python
+class WindowsDataProtectionAgent(object):
+    def __init__(self, **kwargs):
+        # type: (str) -> None
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy'].encode('utf-8')
+            self._entropy_buffer = ctypes.c_buffer(entropy, len(entropy))
+            self._entropy_blob = DATA_BLOB(len(entropy), self._entropy_buffer)
+        else:
+            self._entropy_buffer = None
+            self._entropy_blob = None
+
+    def protect(self, message):
+        # type: (str) -> bytes
+
+        message = message.encode('utf-8')
+        message_buffer = ctypes.c_buffer(message, len(message))
+        message_blob = DATA_BLOB(len(message), message_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_protect_data(
+                ctypes.byref(message_blob),
+                u"python_data",
+                entropy,
+                None,
+                None,
+                _CRYPTPROTECT_UI_FORBIDDEN,
+                ctypes.byref(result)):
+            return result.raw()
+        return b''
+
+    def unprotect(self, cipher_text):
+        # type: (bytes) -> str
+        ct_buffer = ctypes.c_buffer(cipher_text, len(cipher_text))
+        ct_blob = DATA_BLOB(len(cipher_text), ct_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_unprotect_data(
+            ctypes.byref(ct_blob),
+            None,
+            entropy,
+            None,
+            None,
+            _CRYPTPROTECT_UI_FORBIDDEN,
+            ctypes.byref(result)
+        ):
+            return result.raw().decode('utf-8')
+        return u''
+
+
+class _WindowsTokenCache(msal.SerializableTokenCache):
+    DEFAULT_CACHE_LOCATION = os.path.join(os.getenv('LOCALAPPDATA'), '.IdentityService', 'msal.cache')
+    DEFAULT_ENTROPY = ''
+
+    def __init__(self, **kwargs):
+        super(_WindowsTokenCache, self).__init__()
+
+        self._cache_location = _WindowsTokenCache.DEFAULT_CACHE_LOCATION  # type: str
+        if 'cache_location' in kwargs:
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
+        self._lock_location = self._cache_location + '.lockfile'
+
+        entropy = _WindowsTokenCache.DEFAULT_ENTROPY
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy']
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        self._last_sync = 0  # _last_sync is a Unixtime
+
+    def _has_state_changed(self):
+        # type: () -> Bool
+        try:
+            return self.has_state_changed or self._last_sync < os.path.getmtime(self._cache_location)
```

Comment:
Now you might want to skip the `self.has_memory_state_changed` check here, because ideally it shouldn't happen in here in this class.

Reply:
Fixed in 9ef612f